src/MongoDB.Driver.GridFS/GridFSDownloadStreamBase.cs

```diff
@@ -115,8 +124,37 @@ public override Task WriteAsync(byte[] buffer, int offset, int count, Cancellati
         }
 
         // protected methods
+        protected void CloseIfNotAlreadyClosedFromDispose(bool disposing)
+        {
+            if (disposing)
+            {
+                try
+                {
+                    CloseIfNotAlreadyClosed(CancellationToken.None);
+                }
+                catch
+                {
+                    // ignore exceptions when calling CloseIfNotAlreadyClosed from Dispose
+                }
+            }
+        }
+
+        protected virtual void CloseImplementation(CancellationToken cancellationToken)
+        {
+            // override in subclass if there is anything to do
+        }
+
+        protected virtual Task CloseImplementationAsync(CancellationToken cancellationToken)
+        {
+            // override in subclass if there is anything to do
+            CloseImplementation(cancellationToken);
+            return Task.FromResult(true);
+        }
+
         protected override void Dispose(bool disposing)
         {
+            CloseIfNotAlreadyClosedFromDispose(disposing);
+
```

Comment:
If the application called Close explicitly then CloseIfNotAlreadyClosedFromDispose will end up doing nothing.

Otherwise, this is the hook that ensures that no matter how we reached Dispose the Stream will be closed.